libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,476 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+Audacity: A Digital Audio Editor
+
+GetMeterUsingTatumQuantizationFit.cpp
+
+Matthieu Hodgkinson
+
+**********************************************************************/
+
+#include "GetMeterUsingTatumQuantizationFit.h"
+#include "MirAudioReader.h"
+#include "MirDsp.h"
+#include "MirUtils.h"
+#include "MusicInformationRetrieval.h"
+
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <map>
+#include <numeric>
+#include <regex>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace MIR
+{
+namespace
+{
+constexpr auto minTatumsPerMinute = 100;
+constexpr auto maxTatumsPerMinute = 700;
+constexpr auto minBpm = 50.;
+constexpr auto maxBpm = 200.;
+constexpr auto minBeatsPerBar = 2;
+constexpr auto maxBeatsPerBar = 4;
+constexpr std::array<int, 3> possibleBeatsPerBar { 2, 3, 4 };
+constexpr std::array<double, 9> possibleTatumsPerBeat {
+   1.,                             //
+   2. / 1, 3. / 1, 4. / 1, 6. / 1, //
+   1. / 2, 1. / 3, 1. / 4, 1. / 6  //
+};
+
+// Number of bars and beats per bar dividing the input audio recording.
+struct BarDivision
+{
+   const int numBars;
+   const int beatsPerBar;
+};
+
+constexpr bool IsRound(double x)
+{
+   return static_cast<int>(x) == x;
+}
+
+static_assert(IsRound(1.));
+static_assert(!IsRound(2.5));
+
+// A map of possible number of tatums to a list of number of bars and beats per
+// bar which explain it.
+using PossibleDivHierarchies =
+   std::unordered_map<int /*num tatums*/, std::vector<BarDivision>>;
+
+// Gather a collection of possible number of divisions based on the assumption
+// that the audio is a loop (hence there must be a round number of bars) and on
+// reasonable bar and tatum durations.
+PossibleDivHierarchies GetPossibleDivHierarchies(double audioFileDuration)
+{
+   constexpr auto minBarDuration = 1.;
+   constexpr auto maxBarDuration = 4.;
+   const int minNumBars =
+      std::max(std::round(audioFileDuration / maxBarDuration), 1.);
+   const int maxNumBars = std::round(audioFileDuration / minBarDuration);
+   std::vector<int> numBars(maxNumBars - minNumBars + 1);
```

Comment:
I wish we had C++20 ranges.  Without them, you could use standard algorithms without needing these temporary vectors just to hold integers.

Making these vectors seems wasteful and only obscures the two outer loops that might just as easily be written with an integer loop variable.